# A Classification Model for Predicting Whether Individuals got their H1N1 Vaccines

## 1a. Business Understanding

Vaccination has become a key public health measure that is used to fight and in most cases curb infectious diseases. Vaccines provide immunization for individuals and having a community participate in the process can further decrease the spread of diseases as a result of the concept of "herd immunity." 

Our stakeholders in this case are the researchers, the government and patients who hope to benefit from the predictions made by this model. The idea is to come up with a classification modek that will predict (almost) accurately whether individuals were immunized with H1N1 vacciness. 

### b. Problem Statement

### c. Data Understanding

The data used in this project was retrieved from [DrivenData](https://www.drivendata.org/competitions/66/flu-shot-learning/page/210/). The dataset contains data on training features, training labels, test features and the submission format. The data on the training features (the input variables that the model will use to predict the probability that people received H1N1 flu vaccines and seasonal flu vaccines) contains 35 feature columns in total, each a response to a survey question. These questions cover several different topics, such as whether people observed safe behavioral practices, their opinions about the diseases and the vaccines, and their demographics.The data on training labels (the labels corresponding to the observations in the training features) contains two target variables : h1n1_vaccine and seasonal_vaccine. Both are binary variables, with 1 indicating that a person received the respective flu vaccine and 0 indicating that a person did not receive the respective flu vaccine.

### d. metrics of success
Accuracy

Precision

Recall

F1-Score

ROC/AUC curve

## 2. Loading and checking the data

We first start by importing the necessary libaries. 

In [22]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
%matplotlib inline

Next up, Load the dataset

In [2]:
# previewing the top of our dataset
features_df = pd.read_csv("training_set_features.csv")
features_df.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [3]:
# checking the bottom of our dataset
features_df.tail()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26702          26702           2.0             0.0                        0.0   
26703          26703           1.0             2.0                        0.0   
26704          26704           2.0             2.0                        0.0   
26705          26705           1.0             1.0                        0.0   
26706          26706           0.0             0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
26702                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26703                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26704                    1.0  ...                        NaN     Not Married   
26705                    NaN  ...  <= $75,000, Above Poverty         Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   

       rent_or_own   employment_status  hhs_geo_region  \
26702          Own  Not in Labor Force        qufhixun   
26703         Rent            Employed        lzgpxyit   
26704          Own                 NaN        lzgpxyit   
26705         Rent            Employed        lrircsnp   
26706          Own  Not in Labor Force        mlyzmhmf   

                     census_msa  household_adults  household_children  \
26702                   Non-MSA               0.0                 0.0   
26703       MSA, Principle City               1.0                 0.0   
26704  MSA, Not Principle  City               0.0                 0.0   
26705                   Non-MSA               1.0                 0.0   
26706       MSA, Principle City               1.0                 0.0   

       employment_industry  employment_occupation  
26702                  NaN                    NaN  
26703             fcxhlnwr               cmhcxjea  
26704                  NaN                    NaN  
26705             fcxhlnwr               haliazsg  
26706                  NaN                    NaN  

[5 rows x 36 columns]

In [4]:
# checking the number of records in our dataset
features_df.shape

(26707, 36)

This dataset contains 26707 observations and 36 columns. 

The features dataset contains a combination of floats, integers and object data types. 

In [5]:
# Checking whether each column has an appropriate datatype 
features_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   h1n1_concern                 26615 non-null  float64
 2   h1n1_knowledge               26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_h1n1             24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

In [6]:
# checking the top of our labels dataset
labels_df = pd.read_csv("training_set_labels.csv")
labels_df.head()

respondent_id  h1n1_vaccine  seasonal_vaccine
0              0             0                 0
1              1             0                 1
2              2             0                 0
3              3             0                 1
4              4             0                 0

In [7]:
# checking the bottom of our labels dataset
labels_df.tail()

respondent_id  h1n1_vaccine  seasonal_vaccine
26702          26702             0                 0
26703          26703             0                 0
26704          26704             0                 1
26705          26705             0                 0
26706          26706             0                 0

In [8]:
# checking whether the columns have consistent and appropriate data types
labels_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   respondent_id     26707 non-null  int64
 1   h1n1_vaccine      26707 non-null  int64
 2   seasonal_vaccine  26707 non-null  int64
dtypes: int64(3)
memory usage: 626.1 KB


In [9]:
# determining the number of records in our labels dataframe
labels_df.shape

(26707, 3)

The labels dataframe contains 26707 observations and 3 columns.

In [10]:
test_df = pd.read_csv("test_set_features.csv")
test_df.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0          26707           2.0             2.0                        0.0   
1          26708           1.0             1.0                        0.0   
2          26709           2.0             2.0                        0.0   
3          26710           1.0             1.0                        0.0   
4          26711           3.0             1.0                        1.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   1.0                   0.0                    1.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   1.0                    1.0   
3                   0.0                   0.0                    0.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          1.0                      0.0   
1                          0.0                      0.0   
2                          1.0                      1.0   
3                          0.0                      0.0   
4                          1.0                      1.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...                  > $75,000     Not Married   
1                    0.0  ...              Below Poverty     Not Married   
2                    1.0  ...                  > $75,000         Married   
3                    0.0  ...  <= $75,000, Above Poverty         Married   
4                    1.0  ...  <= $75,000, Above Poverty     Not Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0         Rent            Employed        mlyzmhmf  MSA, Not Principle  City   
1         Rent            Employed        bhuqouqj                   Non-MSA   
2          Own            Employed        lrircsnp                   Non-MSA   
3          Own  Not in Labor Force        lrircsnp  MSA, Not Principle  City   
4          Own            Employed        lzgpxyit                   Non-MSA   

   household_adults  household_children  employment_industry  \
0               1.0                 0.0             atmlpfrs   
1               3.0                 0.0             atmlpfrs   
2               1.0                 0.0             nduyfdeo   
3               1.0                 0.0                  NaN   
4               0.0                 1.0             fcxhlnwr   

   employment_occupation  
0               hfxkjkmi  
1               xqwwgdyp  
2               pvmttkik  
3                    NaN  
4               mxkfnird  

[5 rows x 36 columns]

In [11]:
test_df.tail()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
26703          53410           1.0             1.0                        0.0   
26704          53411           3.0             1.0                        0.0   
26705          53412           0.0             1.0                        0.0   
26706          53413           3.0             1.0                        0.0   
26707          53414           2.0             1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
26703                   1.0                   0.0                    1.0   
26704                   1.0                   0.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    1.0   
26707                   0.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
26703                          0.0                      0.0   
26704                          1.0                      1.0   
26705                          0.0                      0.0   
26706                          0.0                      1.0   
26707                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
26703                    1.0  ...                        NaN             NaN   
26704                    1.0  ...              Below Poverty         Married   
26705                    0.0  ...              Below Poverty     Not Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   
26707                    1.0  ...                        NaN     Not Married   

       rent_or_own   employment_status  hhs_geo_region  \
26703          NaN                 NaN        dqpwygqj   
26704         Rent            Employed        qufhixun   
26705         Rent  Not in Labor Force        qufhixun   
26706          Own  Not in Labor Force        bhuqouqj   
26707         Rent            Employed        lrircsnp   

                     census_msa  household_adults  household_children  \
26703       MSA, Principle City               1.0                 1.0   
26704                   Non-MSA               1.0                 3.0   
26705  MSA, Not Principle  City               1.0                 0.0   
26706  MSA, Not Principle  City               1.0                 0.0   
26707       MSA, Principle City               0.0                 0.0   

       employment_industry  employment_occupation  
26703                  NaN                    NaN  
26704             fcxhlnwr               vlluhbov  
26705                  NaN                    NaN  
26706                  NaN                    NaN  
26707                  NaN               xtkaffoo  

[5 rows x 36 columns]

## 4. Data Preparation

Before the data is ready to use, there is need for tidying to ensure consistency, handle missing values, and duplicates.The dataset for one, contains two target variables, the probability that individuals received H1N1 vaccines and seasonal vaccines. For this prject however, the intention is to work with one target variable, the probability of H1N1 vaccine reception, therefore, columns representing the probability of seasonal vaccines will be dropped.

In [12]:
# dropping columns related to seasonal vaccine in the features dataframe
new_features_df = features_df.drop(['doctor_recc_seasonal', 'opinion_seas_vacc_effective', 'opinion_seas_risk', 'opinion_seas_sick_from_vacc'], axis=1)
new_features_df.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children employment_industry  \
0               0.0                 0.0                 NaN   
1               0.0                 0.0            pxcmvdjn   
2               2.0                 0.0            rucpziij   
3               0.0                 0.0                 NaN   
4               1.0                 0.0            wxleyezf   

  employment_occupation  
0                   NaN  
1              xgwztkwe  
2              xtkaffoo  
3                   NaN  
4              emcorrxb  

[5 rows x 32 columns]

In [13]:
# dropping columns related to seasonal vaccine in the labels dataframe
new_labels_df = labels_df.drop(['seasonal_vaccine'], axis=1)
new_labels_df.head()

respondent_id  h1n1_vaccine
0              0             0
1              1             0
2              2             0
3              3             0
4              4             0

In [14]:
test_df

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              26707           2.0             2.0                        0.0   
1              26708           1.0             1.0                        0.0   
2              26709           2.0             2.0                        0.0   
3              26710           1.0             1.0                        0.0   
4              26711           3.0             1.0                        1.0   
...              ...           ...             ...                        ...   
26703          53410           1.0             1.0                        0.0   
26704          53411           3.0             1.0                        0.0   
26705          53412           0.0             1.0                        0.0   
26706          53413           3.0             1.0                        0.0   
26707          53414           2.0             1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                       1.0                   0.0                    1.0   
1                       0.0                   0.0                    0.0   
2                       0.0                   1.0                    1.0   
3                       0.0                   0.0                    0.0   
4                       1.0                   0.0                    1.0   
...                     ...                   ...                    ...   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   0.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    1.0   
26707                   0.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
0                              1.0                      0.0   
1                              0.0                      0.0   
2                              1.0                      1.0   
3                              0.0                      0.0   
4                              1.0                      1.0   
...                            ...                      ...   
26703                          0.0                      0.0   
26704                          1.0                      1.0   
26705                          0.0                      0.0   
26706                          0.0                      1.0   
26707                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
0                        1.0  ...                  > $75,000     Not Married   
1                        0.0  ...              Below Poverty     Not Married   
2                        1.0  ...                  > $75,000         Married   
3                        0.0  ...  <= $75,000, Above Poverty         Married   
4                        1.0  ...  <= $75,000, Above Poverty     Not Married   
...                      ...  ...                        ...             ...   
26703                    1.0  ...                        NaN             NaN   
26704                    1.0  ...              Below Poverty         Married   
26705                    0.0  ...              Below Poverty     Not Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   
26707                    1.0  ...                        NaN     Not Married   

       rent_or_own   employment_status  hhs_geo_region  \
0             Rent            Employed        mlyzmhmf   
1             Rent            Employed        bhuqouqj   
2              Own            Employed        lrircsnp   
3              Own  Not in Labor Force        lrircsnp   
4              Own            Employed        lzgpxyit   
...            ...                 ...             ... 

In [15]:
# dropping columns related to seasonal vaccine in the test features dataframe
new_test_df = test_df.drop(['doctor_recc_seasonal', 'opinion_seas_vacc_effective', 'opinion_seas_risk', 'opinion_seas_sick_from_vacc'], axis=1)
new_test_df

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              26707           2.0             2.0                        0.0   
1              26708           1.0             1.0                        0.0   
2              26709           2.0             2.0                        0.0   
3              26710           1.0             1.0                        0.0   
4              26711           3.0             1.0                        1.0   
...              ...           ...             ...                        ...   
26703          53410           1.0             1.0                        0.0   
26704          53411           3.0             1.0                        0.0   
26705          53412           0.0             1.0                        0.0   
26706          53413           3.0             1.0                        0.0   
26707          53414           2.0             1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                       1.0                   0.0                    1.0   
1                       0.0                   0.0                    0.0   
2                       0.0                   1.0                    1.0   
3                       0.0                   0.0                    0.0   
4                       1.0                   0.0                    1.0   
...                     ...                   ...                    ...   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   0.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    1.0   
26707                   0.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
0                              1.0                      0.0   
1                              0.0                      0.0   
2                              1.0                      1.0   
3                              0.0                      0.0   
4                              1.0                      1.0   
...                            ...                      ...   
26703                          0.0                      0.0   
26704                          1.0                      1.0   
26705                          0.0                      0.0   
26706                          0.0                      1.0   
26707                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
0                        1.0  ...                  > $75,000     Not Married   
1                        0.0  ...              Below Poverty     Not Married   
2                        1.0  ...                  > $75,000         Married   
3                        0.0  ...  <= $75,000, Above Poverty         Married   
4                        1.0  ...  <= $75,000, Above Poverty     Not Married   
...                      ...  ...                        ...             ...   
26703                    1.0  ...                        NaN             NaN   
26704                    1.0  ...              Below Poverty         Married   
26705                    0.0  ...              Below Poverty     Not Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   
26707                    1.0  ...                        NaN     Not Married   

       rent_or_own   employment_status  hhs_geo_region  \
0             Rent            Employed        mlyzmhmf   
1             Rent            Employed        bhuqouqj   
2              Own            Employed        lrircsnp   
3              Own  Not in Labor Force        lrircsnp   
4              Own            Employed        lzgpxyit   
...            ...                 ...             ... 

The resulting dataset has 26708 rows and 32 columns

#### Handling missing values
We check for missing values in the 

In [16]:
# checking for missing values in the dataset
new_features_df.isna().sum()

respondent_id                      0
h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
marital_status                  1408
rent_or_own                     2042
employment_status               1463
hhs_geo_region                     0
c

Following the above outcome, the dataset has a very high number of missing values. To handle missing values, an evaluation is carried out to determine how relevant the rows with the missing values are to the whole process. This determines whether the missing values will be dropped or replaced.

In [17]:
# Check the summary statistics
new_features_df.describe(include = "all")

respondent_id  h1n1_concern  h1n1_knowledge  \
count    26707.000000  26615.000000    26591.000000   
unique            NaN           NaN             NaN   
top               NaN           NaN             NaN   
freq              NaN           NaN             NaN   
mean     13353.000000      1.618486        1.262532   
std       7709.791156      0.910311        0.618149   
min          0.000000      0.000000        0.000000   
25%       6676.500000      1.000000        1.000000   
50%      13353.000000      2.000000        1.000000   
75%      20029.500000      2.000000        2.000000   
max      26706.000000      3.000000        2.000000   

        behavioral_antiviral_meds  behavioral_avoidance  behavioral_face_mask  \
count                26636.000000          26499.000000          26688.000000   
unique                        NaN                   NaN                   NaN   
top                           NaN                   NaN                   NaN   
freq                          NaN                   NaN                   NaN   
mean                     0.048844              0.725612              0.068982   
std                      0.215545              0.446214              0.253429   
min                      0.000000              0.000000              0.000000   
25%                      0.000000              0.000000              0.000000   
50%                      0.000000              1.000000              0.000000   
75%                      0.000000              1.000000              0.000000   
max                      1.000000              1.000000              1.000000   

        behavioral_wash_hands  behavioral_large_gatherings  \
count            26665.000000                  26620.00000   
unique                    NaN                          NaN   
top                       NaN                          NaN   
freq                      NaN                          NaN   
mean                 0.825614                      0.35864   
std                  0.379448                      0.47961   
min                  0.000000                      0.00000   
25%                  1.000000                      0.00000   
50%                  1.000000                      0.00000   
75%                  1.000000                      1.00000   
max                  1.000000                      1.00000   

        behavioral_outside_home  behavioral_touch_face  ...  \
count              26625.000000           26579.000000  ...   
unique                      NaN                    NaN  ...   
top                         NaN                    NaN  ...   
freq                        NaN                    NaN  ...   
mean                   0.337315               0.677264  ...   
std                    0.472802               0.467531  ...   
min                    0.000000               0.000000  ...   
25%                    0.000000               0.000000  ...   
50%                    0.000000               1.000000  ...   
75%                    1.000000               1.000000  ...   
max                    1.000000               1.000000  ...   

                   income_poverty  marital_status  rent_or_own  \
count                       22284           25299        24665   
unique                          3               2            2   
top     <= $75,000, Above Poverty         Married          Own   
freq                        12777           13555        18736   
mean                          NaN             NaN          NaN   
std                           NaN             NaN          NaN   
min                           NaN             NaN          NaN   
25%                           NaN             NaN          NaN   
50%                           NaN             NaN          NaN   
75%                           NaN             NaN          NaN   
max                           NaN             NaN          NaN   

        employment_status  hhs_geo_region                census_msa  \
count   

It is safe to say that the dataset will still be of great use and quality without the rows that contain missing values. For this reason, we drop them

In [18]:
# dropping all the rows with missing values
new_features_df = new_features_df.dropna()
new_features_df

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
1                  1           3.0             2.0                        0.0   
7                  7           1.0             0.0                        0.0   
10                10           2.0             1.0                        0.0   
11                11           1.0             2.0                        0.0   
15                15           1.0             1.0                        0.0   
...              ...           ...             ...                        ...   
26681          26681           2.0             1.0                        0.0   
26683          26683           1.0             2.0                        1.0   
26690          26690           0.0             2.0                        0.0   
26699          26699           2.0             2.0                        0.0   
26703          26703           1.0             2.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
1                       1.0                   0.0                    1.0   
7                       1.0                   0.0                    1.0   
10                      1.0                   0.0                    1.0   
11                      1.0                   0.0                    1.0   
15                      1.0                   0.0                    0.0   
...                     ...                   ...                    ...   
26681                   1.0                   0.0                    1.0   
26683                   0.0                   1.0                    1.0   
26690                   0.0                   0.0                    1.0   
26699                   1.0                   0.0                    1.0   
26703                   1.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
1                              0.0                      1.0   
7                              0.0                      0.0   
10                             1.0                      0.0   
11                             0.0                      0.0   
15                             0.0                      0.0   
...                            ...                      ...   
26681                          0.0                      0.0   
26683                          1.0                      0.0   
26690                          0.0                      0.0   
26699                          1.0                      0.0   
26703                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
1                        1.0  ...              Below Poverty     Not Married   
7                        1.0  ...  <= $75,000, Above Poverty         Married   
10                       0.0  ...  <= $75,000, Above Poverty         Married   
11                       0.0  ...                  > $75,000         Married   
15                       0.0  ...                  > $75,000         Married   
...                      ...  ...                        ...             ...   
26681                    1.0  ...  <= $75,000, Above Poverty         Married   
26683                    1.0  ...                  > $75,000         Married   
26690                    1.0  ...  <= $75,000, Above Poverty         Married   
26699                    1.0  ...  <= $75,000, Above Poverty         Married   
26703                    0.0  ...  <= $75,000, Above Poverty     Not Married   

       rent_or_own  employment_status  hhs_geo_region  \
1             Rent           Employed        bhuqouqj   
7              Own           Employed        bhuqouqj   
10            Rent           Employed        lzgpxyit   
11             Own           Employed        oxchjgsf   
15             Own           Employed        lzgpxyit   
...            ...                ...             ...   
2668

In [20]:
# we check the dataset again to ensure we have handled all missing values
new_features_df.isna().sum()

respondent_id                  0
h1n1_concern                   0
h1n1_knowledge                 0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_h1n1               0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_h1n1_vacc_effective    0
opinion_h1n1_risk              0
opinion_h1n1_sick_from_vacc    0
age_group                      0
education                      0
race                           0
sex                            0
income_poverty                 0
marital_status                 0
rent_or_own                    0
employment_status              0
hhs_geo_region                 0
census_msa                     0
household_adults               0
household_children             0
employment

The dataset does not have anymore missing values

#### cheking for duplicates

In [21]:
# check for duplicates
new_features_df.duplicated().any()

False

The dataset has no duplicates

## 5. Exploratory Data Analysis